### ISSUE TITLE
SF does not provide unique port for multiple services of same service type

### ISSUE THREAD

**Issue by Vhab** [Azure/service-fabric-issues#83]
tl;dr: Multiple services of same service type do not get unique ports assigned.

A bit of background:

We're launching multiple different named services of a single service type with different launch configuration.
Eg: `SomeServiceType` is launched as `SomeService/One` and `SomeService/Two`, each with their own `InstanceCount`.
The combined InstanceCount of these services will exceed the amount of Nodes in our cluster.

`SomeServiceType` has an endpoint resource in `ServiceManifest.xml` without a port specified, eg: `<Endpoint Protocol="tcp" Name="SomeEndpoint" Type="Input" />`.
SF will correct assign unique port numbers to services when `SomeServiceType` is launched once.

However here's our issue: when a second service is started, `SomeService/One`'s assigned ports will collide with `SomeService/Two`.

This makes us assume SF assigns unique ports based on a combination of Node+ServiceType instead of the expected Node+Service.

It is critical for us to be able to run multiple Services of the same ServiceType on a single cluster. We're open to suggestions for workarounds, but would prefer to see a fix in SF itself.

**Comment by masnider** [comment]
TL;DR - you probably should just create more application instances

This is currently by design since we strongly recommend that services use a transport that supports port reuse and disambiguation on other aspects of the addressing scheme, since this allows greater density. A lot of transports support this (notably HTTP at least as delivered through http.sys on Windows, and it can be achieved through things like TCP port sharing), though some don't or require additional work by you the developer to do so. If your transport doesn't allow this or it is arduous, then you have the following options, all of which are fully supported (though some are a bit ugly):

- Define an endpoint per service in your service manifest. This is a solution since we will hand out different ports per endpoint, however we understand that it is a bit ugly and does set up an additional management burden (as well as being a bit of a bad pattern since it couples the instance information [endpoint names] to the service type definition [the manifest] )
- Create different application instances for every group of services that you want to end up in different processes. By default Service Fabric scopes the port allocations per-endpoint-per-process (and process allocations are per application instance). This should give you what you want and is probably the best answer. In the future we're considering adding density mode options when you create applications and services in order to give you direct control over this scoping. 
- You can actually choose not to use our endpoint naming and setup definitions at all. The downside to this is that you will have to take care of being sure that your service can acl and secure the ports within its own code. _Sometimes_ (again depending on transport and other things) this requires the service to run at a higher permissions level than you'd really want - so you may have to introduce another service to do the local port management, etc. Not ideal, and a little complicated, but always a way out. Service Fabric will always take the string address and endpoint name returned from a communication listener and use it for discovery, so you have full control.

HTH

**Comment by Vhab** [comment]
Thank you for your detailed reply.

Unfortunately we're unable to select a transport that supports port sharing for our specific needs. We're fighting a combination of specific routing and low latency needs, which lead us to use some custom-ish solutions (we're aware we go against the grain of SF here).

Your suggestions have given us some good food for thought for a work around, we'll pursue one of these options.
The last option looks closest to tooling we've already build, as we're already dynamically spinning new services up and down.

That said, we are happy with SF's ability to let services report back their endpoints and being able to query this data on the cluster. This is making a workaround much more feasible.

Should you consider it in the future, we'd love a way to tell SF at runtime we want a unique port and take benefit of SF's ACL management.
Being able to make a service request a new non-colliding endpoint at runtime, or being able to specify endpoint needs in `ServiceDescription` before feeding it to `CreateServiceAsync` would very useful to our use-case.

**Comment by masnider** [comment]
Tagged as such. We have some thoughts along these lines and you should see this getting easier in future releases. If you have any further issues getting this to work let us know.

**Comment by fizxmike** [comment]
Any update on this? The use case for me is *Code Reuse*... Specifically for guest executable (where I don't think I have access to "ConfigPackages." Since many of my services use the very same large set of libraries (100s of MB), with each service (entry-point) essentially being a unique implementation of business logic on top, I'd like to not have to COPY THE ENTIRE SET OF LIBRARIES for each service at package build time.

I guess what I'm looking for is the docker-compose analog, where you can re-use container images then set environment variables and entry-point commands for each service instance... Maybe I do Application Instances instead... but then my "service collection" is defined inside a PowerShell script, SMH...

**Comment by fizxmike** [comment]
So I tried defining multiple application instances, but my use case also includes overriding ports (legacy service written in java). The resource override feature is broken for me: https://github.com/Azure/service-fabric-issues/issues/447 https://github.com/MicrosoftDocs/azure-docs/issues/5505
